# Analysis of Ebay Car Sales Data
Goal: clean and analyse the included used car listings. 
# Dataset
I will work on the autos.csv data set to perform my analysis. This is a data set of used cars from eBay Kleinanzeigen.

| Columns            | Definition                  |
| ------------------ | --------------------------- |
| dateCrawled        | When ad was first crawled   |
| name               | Name of the car             |
| seller             | Seller private or dealer    |
| offerType          | The type of listing         |
| price              | Price on ad to sell the car |
| abtest             | Listing included in A/B test|
| vechileType        | The vehicle type            |
| yearOfRegistration | Year car first registered   |
| gearbox            | The transmission type       |
| powerPS            | Power of the car in PS      |
| model              | Car model name              |
| kilometer          | How many kms car has driven |
| monthOfRegistration| Month car first registered  |
| fuelType           | Type of fuel the car uses   |
| brand              | Brand of the car            |
| notRepairedDamage  | Car has damage not repaired |
| dateCreated        | Date eBay listing created   |
| nrOfPictures       | Number of pictures in ad    |
| postalCode         | Postal code for location    |
| lastSeenOnline     | Crawler last saw ad online  |

In [2]:
import pandas as pd

In [4]:
import numpy as np

Importing data set

In [10]:
autos = pd.read_csv("C:\\Users\\eliza\\OneDrive\\Documents\\Grayce\\Training\\Guided_Project3\\data-society-used-cars-data\\autos.csv", encoding="Latin-1")

# Initial Analysis of Data Set

In [13]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

In [14]:
autos.describe()

price  yearOfRegistration        powerPS      kilometer  \
count  3.715280e+05       371528.000000  371528.000000  371528.000000   
mean   1.729514e+04         2004.577997     115.549477  125618.688228   
std    3.587954e+06           92.866598     192.139578   40112.337051   
min    0.000000e+00         1000.000000       0.000000    5000.000000   
25%    1.150000e+03         1999.000000      70.000000  125000.000000   
50%    2.950000e+03         2003.000000     105.000000  150000.000000   
75%    7.200000e+03         2008.000000     150.000000  150000.000000   
max    2.147484e+09         9999.000000   20000.000000  150000.000000   

       monthOfRegistration  nrOfPictures    postalCode  
count        371528.000000      371528.0  371528.00000  
mean              5.734445           0.0   50820.66764  
std               3.712412           0.0   25799.08247  
min               0.000000           0.0    1067.00000  
25%               3.000000           0.0   30459.00000  
50%               6.000000           0.0   49610.00000  
75%               9.000000           0.0   71546.00000  
max              12.000000           0.0   99998.00000

In [15]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

From initial analysis of the data set, vehicleType, gearbox, model, fuelType, notRepairedDamage all contain null values. These will need to be addressed. Seven columns are int64 type and the other 13 columns are object type. There are 20 columns in total. The minimum price is £0.00, which may need to be investigated.  

# Changing column names

Existing columns

In [17]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Updating column names

In [30]:
import stringcase

column_name_list = []

for a in autos.columns:
    a = a.replace('yearOfRegistration','registration_year')
    a = a.replace('monthOfRegistration','registration_month')
    a = a.replace('notRepairedDamage','unrepaired_damage')
    a = a.replace('dateCreated','ad_created')
    # Converting from camelcase to snakecase
    new_string = stringcase.snakecase(a)
    column_name_list.append(new_string)
    
autos.columns = column_name_list

In [32]:
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_p_s  model  \
0    480   test          NaN               1993    manuell          0   golf   
1  18300   test        coupe               2011    manuell        190    NaN   
2   9800   test          suv               2004  automatik        163  grand   
3   1500   test   kleinwagen               2001    manuell         75   golf   
4   3600   test   kleinwagen               2008    manuell         69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

I have changed the names of the columns to give a better description of their contents.

# Cleaning data further

In [35]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox      power_p_s  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-06 13:45:54  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

In [36]:
autos.describe()

price  registration_year      power_p_s      kilometer  \
count  3.715280e+05      371528.000000  371528.000000  371528.000000   
mean   1.729514e+04        2004.577997     115.549477  125618.688228   
std    3.587954e+06          92.866598     192.139578   40112.337051   
min    0.000000e+00        1000.000000       0.000000    5000.000000   
25%    1.150000e+03        1999.000000      70.000000  125000.000000   
50%    2.950000e+03        2003.000000     105.000000  150000.000000   
75%    7.200000e+03        2008.000000     150.000000  150000.000000   
max    2.147484e+09        9999.000000   20000.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count       371528.000000        371528.0  371528.00000  
mean             5.734445             0.0   50820.66764  
std              3.712412             0.0   25799.08247  
min              0.000000             0.0    1067.00000  
25%              3.000000             0.0   30459.00000  
50%              6.000000             0.0   49610.00000  
75%              9.000000             0.0   71546.00000  
max             12.000000             0.0   99998.00000

I can drop columns that have mostly one value, these include `seller` and `offer_type`. `nr_of_pictures` just seems to have the value `0` so can also be dropped.

In [45]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                  int64
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_p_s              int64
model                 object
kilometer              int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
nr_of_pictures         int64
postal_code            int64
last_seen             object
dtype: object

# Analysing `kilometer` and `price` columns

Starting with the `kilometer` column:

In [48]:
autos["kilometer"].unique() # Lists the unique values

array([150000, 125000,  90000,  40000,  30000,  70000,   5000, 100000,
        60000,  20000,  80000,  50000,  10000], dtype=int64)

In [50]:
autos["kilometer"].unique().shape # There are 13 unique values

(13,)

In [51]:
autos["kilometer"].describe()

count    371528.000000
mean     125618.688228
std       40112.337051
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: kilometer, dtype: float64

In [57]:
autos["kilometer"].value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: kilometer, dtype: int64

The column `kilometer` has reasonable values, even though above the 50% percentile is 150000.

Looking at the `price` column:

In [59]:
autos["price"].unique()

array([  480, 18300,  9800, ..., 18429, 24895, 10985], dtype=int64)

In [61]:
autos["price"].unique().shape

(5597,)

In [62]:
autos["price"].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [63]:
autos["price"].value_counts()

0         10778
500        5670
1500       5394
1000       4649
1200       4594
          ...  
23456         1
171000        1
21830         1
13485         1
8188          1
Name: price, Length: 5597, dtype: int64

Looking at the least common 20 prices

In [68]:
autos["price"].value_counts(ascending=True).head(20)

8188      1
13485     1
21830     1
171000    1
23456     1
30530     1
42840     1
12973     1
2730      1
145900    1
8265      1
77        1
10670     1
4175      1
28995     1
14666     1
19269     1
3755      1
31555     1
588       1
Name: price, dtype: int64

Looking at the most common 20 prices

In [70]:
autos["price"].value_counts(ascending=False).head(20)

0       10778
500      5670
1500     5394
1000     4649
1200     4594
2500     4438
600      3819
3500     3792
800      3784
2000     3432
999      3364
750      3203
650      3150
4500     3053
850      2946
2200     2936
700      2936
1800     2886
900      2874
950      2793
Name: price, dtype: int64

`Price` should not be `$0`, so I will remove all 10778 entries with a price of `$0`

In [78]:
autos.sort_values(by=['price']).tail(50)

date_crawled  \
91106   2016-03-30 19:46:05   
54295   2016-03-08 11:51:40   
176070  2016-04-02 14:43:10   
281833  2016-03-19 08:50:50   
358868  2016-03-16 16:49:53   
363434  2016-03-23 06:36:20   
121793  2016-03-05 20:59:54   
171824  2016-03-29 23:52:57   
196306  2016-04-01 01:59:20   
361451  2016-04-03 14:36:47   
14663   2016-04-03 06:03:56   
127769  2016-03-28 12:51:51   
185227  2016-03-14 16:52:51   
318143  2016-03-21 13:57:51   
90853   2016-04-01 05:36:16   
54507   2016-03-09 11:40:15   
177744  2016-03-31 16:37:00   
210657  2016-03-12 21:50:57   
338415  2016-03-26 22:49:39   
332028  2016-03-31 18:39:53   
287835  2016-03-26 00:56:36   
266124  2016-04-03 23:36:16   
241847  2016-03-31 18:56:54   
310924  2016-03-14 05:36:22   
39625   2016-03-08 23:53:51   
217547  2016-03-11 10:40:52   
299872  2016-03-14 10:06:23   
20746   2016-03-08 15:54:52   
273145  2016-03-09 15:45:47   
279937  2016-04-03 11:56:00   
87799   2016-03-08 20:39:05   
183361  2016-03-26 11:51:20   
120919  2016-03-25 12:43:02   
251980  2016-03-21 22:25:19   
362782  2016-03-06 10:38:42   
150909  2016-04-02 22:41:54   
246635  2016-03-25 11:55:36   
108987  2016-03-29 00:38:37   
127820  2016-03-14 17:49:54   
280059  2016-03-05 20:49:28   
69747   2016-03-20 10:59:32   
77520   2016-03-22 14:58:27   
139052  2016-03-14 12:49:48   
268173  2016-04-01 09:37:19   
325589  2016-03-27 21:55:14   
56973   2016-03-18 18:52:27   
160530  2016-03-29 12:58:27   
366653  2016-03-20 10:59:52   
189737  2016-04-03 12:41:33   
129969  2016-03-11 08:37:57   

                                                     name  seller offer_type  \
91106   Verkaufe_oder_tausche_wenn_nur_gegen_Kombi_s4_...  privat    Angebot   
54295                        Mazda_5_2.0_CD_DPF_Exclusive  privat    Angebot   
176070                                      Suche_BMW_Fun  privat    Angebot   
281833              Coole_Autowerkstatt_in_Esens_gesucht!  privat    Angebot   
358868                     Tausche_BMW_E46_Touring_Diesel  privat    Angebot   
363434                     Vw_Passat_35i_1.8_schlachtfest  privat    Angebot   
121793                       Alter_VW_Kaefer_Scheunenfund  privat    Angebot   
171824  suche_maserati_3200_gt_Zustand_unwichtig_laufe...  privat    Angebot   
196306                                Suche_Motorcross_.!  privat    Angebot   
361451                                  Hilfe_Hilfe_Hilfe  privat    Angebot   
14663                          Tausch_moeglich_Polo_vw_n9  privat    Angebot   
127769                   Golf_4_laesst_sich_nicht_oeffnen  privat    Angebot   
185227                               Suche_fuer_mein_golf  privat    Angebot   
318143                            Schlachte_Golf_3_gt_tdi  privat    Angebot   
90853                               Verkaufe_oder_Tausche  privat    Angebot   
54507                    Tausch_DKW_F8_gegen_Lanz_Bulldog  privat    Angebot   
177744                     Suche_Trabant_auf_Ratenzahlung  privat    Angebot   
210657   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000  privat    Angebot   
338415                         Polo_6n_gti__Zustand_egal_  privat    Angebot   
332028                     Golf_3_Gti_"16v_Turbo_vr6_bbs"  privat    Angebot   
287835                                 tausche_o_verkaufe  privat    Angebot   
266124  Audi_A4_b7_8e_no_airride_static_20_Zoll_nur_Ta...  privat    Angebot   
241847  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...  privat    Angebot   
310924                     Wer_will_mich_kleinen_polo_6n1  privat    Angebot   
39625                         Tausche_volvo_v40_gegen_van  privat    Angebot   
217547                                         Fiat_Punto  privat    Angebot   
299872                          Vw_polo_6n_Colour_Concept  privat    Angebot   
20746                                  Golf3_Schlachtfest  privat    Angebot   
273145                                         Fiat_Punto  privat    Angebot   
279937  Mercedes_Benz_CLK_Coupe_2

The car prices seem much too high. I will just keep price for between `$1` and `$400,000`

In [81]:
autos = autos[(autos["price"] >= 1) & (autos["price"] <= 400000)]

# Exploring the date columns

These columns include: 
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Currently, `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Let us look at how we can make use of these columns better for our analysis.

In [84]:
autos[["date_crawled","last_seen","ad_created"]].head(10)

date_crawled            last_seen           ad_created
0   2016-03-24 11:52:17  2016-04-07 03:16:57  2016-03-24 00:00:00
1   2016-03-24 10:58:45  2016-04-07 01:46:50  2016-03-24 00:00:00
2   2016-03-14 12:52:21  2016-04-05 12:47:46  2016-03-14 00:00:00
3   2016-03-17 16:54:04  2016-03-17 17:40:17  2016-03-17 00:00:00
4   2016-03-31 17:25:20  2016-04-06 10:17:21  2016-03-31 00:00:00
5   2016-04-04 17:36:23  2016-04-06 19:17:07  2016-04-04 00:00:00
6   2016-04-01 20:48:51  2016-04-05 18:18:39  2016-04-01 00:00:00
8   2016-04-04 23:42:13  2016-04-04 23:42:13  2016-04-04 00:00:00
9   2016-03-17 10:53:50  2016-03-31 17:17:06  2016-03-17 00:00:00
10  2016-03-26 19:54:18  2016-04-06 10:45:34  2016-03-26 00:00:00

In [86]:
# Reassign the cleaned values back to the columns
autos["date_crawled"]=autos["date_crawled"].str[:10]
autos["last_seen"]=autos["last_seen"].str[:10]
autos["ad_created"]=autos["ad_created"].str[:10]

In [98]:
# Further analysis of date_crawled column
autos["date_crawled"].value_counts(normalize=True).sort_index(ascending=True)

2016-03-05    0.025546
2016-03-06    0.014483
2016-03-07    0.035656
2016-03-08    0.033468
2016-03-09    0.034114
2016-03-10    0.032647
2016-03-11    0.032772
2016-03-12    0.036241
2016-03-13    0.015783
2016-03-14    0.036330
2016-03-15    0.033424
2016-03-16    0.030205
2016-03-17    0.031649
2016-03-18    0.013118
2016-03-19    0.035271
2016-03-20    0.036399
2016-03-21    0.035681
2016-03-22    0.032492
2016-03-23    0.032004
2016-03-24    0.029913
2016-03-25    0.032800
2016-03-26    0.031974
2016-03-27    0.030227
2016-03-28    0.035063
2016-03-29    0.034125
2016-03-30    0.033535
2016-03-31    0.031874
2016-04-01    0.034145
2016-04-02    0.035096
2016-04-03    0.038811
2016-04-04    0.037627
2016-04-05    0.012783
2016-04-06    0.003128
2016-04-07    0.001617
Name: date_crawled, dtype: float64

In [97]:
# Further analysis of last_seen column
autos["last_seen"].value_counts(normalize=True).sort_index(ascending=True)

2016-03-05    0.001264
2016-03-06    0.004098
2016-03-07    0.005202
2016-03-08    0.007939
2016-03-09    0.009824
2016-03-10    0.011460
2016-03-11    0.012955
2016-03-12    0.023239
2016-03-13    0.008410
2016-03-14    0.012176
2016-03-15    0.016324
2016-03-16    0.016418
2016-03-17    0.028702
2016-03-18    0.006888
2016-03-19    0.016329
2016-03-20    0.019884
2016-03-21    0.020025
2016-03-22    0.020508
2016-03-23    0.018015
2016-03-24    0.019163
2016-03-25    0.019000
2016-03-26    0.015958
2016-03-27    0.016720
2016-03-28    0.022188
2016-03-29    0.023284
2016-03-30    0.023724
2016-03-31    0.024243
2016-04-01    0.023896
2016-04-02    0.024967
2016-04-03    0.025310
2016-04-04    0.025535
2016-04-05    0.126968
2016-04-06    0.218949
2016-04-07    0.130434
Name: last_seen, dtype: float64

In [96]:
# Further analysis of ad_created column
autos["ad_created"].value_counts(normalize=True).sort_index(ascending=True)

2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.039000
2016-04-04    0.037736
2016-04-05    0.011615
2016-04-06    0.003119
2016-04-07    0.001553
Name: ad_created, Length: 114, dtype: float64

`last_seen` and `date_crawled` columns are similar. It was more popular to create ads in 2016.

# Registration year

In [100]:
autos["registration_year"].describe()

count    360641.000000
mean       2004.433023
std          81.016330
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

The minimum registration year is `1000`, before cars were invented. The maximum registration year is `9999`, which hasn't happened yet. So neither of these are valid. 

I have decided to only include cars from the 1900-2016. 

In [101]:
autos = autos[(autos["registration_year"] >= 1900) & (autos["registration_year"] <= 2016)]

In [105]:
autos["registration_year"].value_counts(normalize=True)

2000    0.066698
1999    0.063551
2005    0.062668
2006    0.057707
2001    0.056957
          ...   
1928    0.000006
1927    0.000006
1940    0.000006
1925    0.000003
1911    0.000003
Name: registration_year, Length: 94, dtype: float64

More cars are made recently, with fewer registered back to the early 1900s.

# Exploring Price by Brand

In [107]:
autos["brand"].unique()

array(['volkswagen', 'audi', 'jeep', 'skoda', 'bmw', 'peugeot', 'ford',
       'mazda', 'nissan', 'renault', 'mercedes_benz', 'seat', 'honda',
       'fiat', 'opel', 'mini', 'smart', 'hyundai', 'sonstige_autos',
       'alfa_romeo', 'subaru', 'volvo', 'mitsubishi', 'kia', 'suzuki',
       'lancia', 'porsche', 'citroen', 'toyota', 'chevrolet', 'dacia',
       'daihatsu', 'trabant', 'chrysler', 'jaguar', 'daewoo', 'rover',
       'saab', 'land_rover', 'lada'], dtype=object)

In [109]:
autos["brand"].value_counts()

volkswagen        73388
bmw               38089
opel              36888
mercedes_benz     33572
audi              31041
ford              23891
renault           16472
peugeot           10453
fiat               8906
seat               6469
skoda              5438
mazda              5333
smart              4968
citroen            4836
nissan             4714
toyota             4483
hyundai            3457
sonstige_autos     3293
mini               3253
volvo              3171
mitsubishi         2855
honda              2611
kia                2397
suzuki             2206
alfa_romeo         2187
porsche            2155
chevrolet          1741
chrysler           1339
dacia               865
jeep                760
land_rover          751
daihatsu            749
subaru              734
jaguar              601
saab                508
daewoo              505
trabant             488
lancia              451
rover               441
lada                207
Name: brand, dtype: int64

I have chosen to aggregate on the top 10 brands.

In [116]:
selected_brands= autos["brand"].value_counts().head(10).index

In [130]:
brand_mean_price = {}
brand_mean_mileage = {}
for a in selected_brands:
    brand_rows = autos[autos["brand"] == a]
    mean_price = brand_rows["price"].mean() # Calculating the mean price for this subset of the data
    mean_mileage = brand_rows["kilometer"].mean() # Calculating the mean milage for this subset of the data
    brand_mean_price[a] = round(mean_price, 2)
    brand_mean_mileage[a] = round(mean_mileage, 2)

Out of the ten most common cars, Audi is the most expensive on average, and Renault is the least expensive on average.

Converting both dictionaries to series objects using the series constructor.

In [131]:
bmp_series = pd.Series(brand_mean_price)
bmm_series = pd.Series(brand_mean_mileage)

In [133]:
print(bmp_series)

volkswagen       5400.19
bmw              8459.14
opel             2971.90
mercedes_benz    8562.36
audi             9086.28
ford             3696.41
renault          2437.61
peugeot          3267.61
fiat             2892.95
seat             4542.00
dtype: float64


In [135]:
print(bmm_series)

volkswagen       128386.04
bmw              132796.74
opel             128722.62
mercedes_benz    130569.97
audi             129443.96
ford             123662.26
renault          127885.20
peugeot          124500.62
fiat             116619.69
seat             120608.29
dtype: float64


Creating a dataframe

In [137]:
df = pd.DataFrame(bmp_series, columns=["mean_price"])

In [139]:
df

mean_price
volkswagen        5400.19
bmw               8459.14
opel              2971.90
mercedes_benz     8562.36
audi              9086.28
ford              3696.41
renault           2437.61
peugeot           3267.61
fiat              2892.95
seat              4542.00

Adding the other column

In [141]:
df["mean_mileage"] = bmm_series

In [142]:
df

mean_price  mean_mileage
volkswagen        5400.19     128386.04
bmw               8459.14     132796.74
opel              2971.90     128722.62
mercedes_benz     8562.36     130569.97
audi              9086.28     129443.96
ford              3696.41     123662.26
renault           2437.61     127885.20
peugeot           3267.61     124500.62
fiat              2892.95     116619.69
seat              4542.00     120608.29

The average mileage does not strongly influence the mean price of the car. The more expensive cars have a slightly higher mean mileage than the others, but it is not significant. 

In [143]:
df["mean_mileage"].describe()

count        10.000000
mean     126319.539000
std        4936.866481
min      116619.690000
25%      123871.850000
50%      128135.620000
75%      129263.625000
max      132796.740000
Name: mean_mileage, dtype: float64